### ISSUE TITLE
TapTree roundtrip serialization fails with hidden branches

### ISSUE THREAD

**Issue by dr-orlovsky** [rust-bitcoin/rust-bitcoin#928]
`TapTree` can be build from `TaprootBuilder` having hidden branches; however PSBT does not provide format for presenting hidden branches. Thus, `TapTree` psbt serialization is buggy: it will present incorrect representation of trees/descriptors with hidden branches. As a result, `TapTree` also fails to roundtrip if a hidden branch is present:

```rust
    #[test]
    fn taptree_roundtrip_hidden() {
        let mut builder = compose_taproot_builder(0x51, [2, 2, 2]);
        builder = builder.add_leaf_with_ver(3, Script::from_hex("b9").unwrap(), LeafVersion::from_consensus(0xC2).unwrap()).unwrap();
        builder = builder.add_hidden(3, sha256::Hash::default()).unwrap();
        let tree = TapTree::from_inner(builder).unwrap();
        let tree_prime = TapTree::deserialize(&tree.serialize()).unwrap();
        assert_eq!(tree, tree_prime);
    }
```
fails: `test result: FAILED. 248 passed; 1 failed`

---- util::psbt::serialize::tests::taptree_roundtrip_hidden stdout ----
thread 'util::psbt::serialize::tests::taptree_roundtrip_hidden' panicked at 'called `Result::unwrap()` on an `Err` value: ParseFailed("Tree not in DFS order")', src/util/psbt/serialize.rs:415:66

The proposed solution is to introduce `TaprootBuilder::has_hidden_branches` and check that in `TapTree::from_inner`, returning error. I am already working on a PR fixing that.

Related discussion: https://github.com/rust-bitcoin/rust-bitcoin/pull/924#issuecomment-1085387179

**Comment by dr-orlovsky** [comment]
Sorry for misunderstanding the error message: this is not a bug since the presence of the hidden branches is checked in `TaprootBuilder::is_complete` and `TapTree::from_inner` fails with weird `Err(TaprootBuilder)` error. Thus, it is not the roundtrip serialization which fails, but the tap tree construction.

PS. I think we should return a error type, not a builder; otherwise this is really confusing.

**Comment by dr-orlovsky** [comment]
Converted from "bug" to a "question", which is actually "should we change the return type of `TapTree::from_inner` to a proper error"

**Comment by sanket1729** [comment]
> his is not a bug since the presence of the hidden branches is checked in TaprootBuilder::is_complete

We only check whether the tree is complete. We don't check if the tree has no hidden nodes.

**Comment by sanket1729** [comment]
> The proposed solution is to introduce TaprootBuilder::has_hidden_branches and check that in TapTree::from_inner, returning error. I am already working on a PR fixing that.

Indeed, we should do something like this. I thought I implemented it in my original suggestion, but looks like I forgot to do so :P

**Comment by sanket1729** [comment]
From @Kixunil (I don't know why github links are not working)

AFAIK it's a [convention](https://doc.rust-lang.org/stable/std/string/struct.String.html#method.from_utf8) in Rust to not destroy data like this. Result<Self, TaprootBuilder> would be better.

**Comment by dr-orlovsky** [comment]
Sorry for all the confusion. Yes, the complete tree with hidden branches can be converted into `TapTree`, which will fail serialization roundtrip. Fixed the test in the error description and test output.

**Comment by dr-orlovsky** [comment]
@sanket1729 added fix to the #929

**Comment by apoelstra** [comment]
I'm sure achow is sick of us by now, but IMO this seems like a missing feature in PSBT.
